packages/datadog-instrumentations/src/mocha.js

```diff
@@ -5,101 +5,117 @@ const testStartCh = channel('ci:mocha:test:start')
 const errorCh = channel('ci:mocha:test:error')
 const skipCh = channel('ci:mocha:test:skip')
 const testFinishCh = channel('ci:mocha:test:finish')
-const suiteFinishCh = channel('ci:mocha:suite:finish')
-const hookErrorCh = channel('ci:mocha:hook:error')
 const parameterizedTestCh = channel('ci:mocha:test:parameterize')
 const testRunFinishCh = channel('ci:mocha:run:finish')
 
-// TODO: remove when root hooks and fixtures are implemented
-const patched = new WeakSet()
-
 function isRetry (test) {
   return test._currentRetry !== undefined && test._currentRetry !== 0
 }
 
-function getAllTestsInSuite (root) {
-  const tests = []
-  function getTests (suiteOrTest) {
-    suiteOrTest.tests.forEach(test => {
-      tests.push(test)
-    })
-    suiteOrTest.suites.forEach(suite => {
-      getTests(suite)
-    })
-  }
-  getTests(root)
-  return tests
-}
+// TODO: remove when root hooks and fixtures are implemented
+const patched = new WeakSet()
+
+const testToAr = new WeakMap()
+const originalFns = new WeakMap()
 
 function mochaHook (Runner) {
   if (patched.has(Runner)) return Runner
 
   patched.add(Runner)
 
-  shimmer.wrap(Runner.prototype, 'runTest', runTest => function () {
-    if (!testStartCh.hasSubscribers || isRetry(this.test)) {
-      return runTest.apply(this, arguments)
+  shimmer.wrap(Runner.prototype, 'run', run => function () {
+    if (!testStartCh.hasSubscribers) {
```

Comment:
run is only called once per test run, so we set up here every listener that we need